packages/sources/layer2-sequencer-health/src/starkware.ts

```diff
@@ -0,0 +1,118 @@
+        entrypoint: 'initialize',
+        calldata: [starkKeyPub, '0'],
+      },
+      undefined,
+      { maxFee: '0' },
+    ),
+    error: `Transaction receipt not received in ${config.timeoutLimit} milliseconds`,
+  })
+  Logger.info(
+    `Transaction receipt received with hash ${receipt.transaction_hash} for EVM network: ${Networks.Starkware}`,
+  )
+  await provider.waitForTransaction(receipt.transaction_hash)
+  const tx = await provider.getTransaction(receipt.transaction_hash)
+  return tx.status !== 'REJECTED'
+}
+
+/**
+ * The centralized Starkware sequencer is made up of two internal components.  The first is the gateway which is used to query the Sequencer state
+ * and the batcher which adds transactions to the pending block and processes the pending block to add it into the L2 chain.  The
+ * Sequencer is considered unhealthy if either of the components are down.
+ *
+ *  - The gateway is considered to be down if request to fetch pending block times out or if we get a 404
+ *  - The batcher is considered to be down if we a new block has not been produced AND no new transactions have been added to the
+ *  pending block within a given time interval.
+ * status when fetching
+ * @param networks
+ * @returns true if Sequencer is healthy
+ */
+export const checkStarkwareSequencerPendingTransactions = (): ((
+  config: ExtendedConfig,
+) => Promise<boolean>) => {
+  let lastSeenPendingBlock: PendingBlockState = {
+    parentBlockHash: '',
+    txnCount: 0,
+    lastUpdated: 0,
+    lastStatus: true,
+  }
+  return async (config: ExtendedConfig): Promise<boolean> => {
+    const delta = config.delta
+    const currentTime = Date.now()
+    if (
+      lastSeenPendingBlock.lastUpdated > 0 &&
+      currentTime - lastSeenPendingBlock.lastUpdated < delta
+    ) {
+      Logger.debug(
+        `Skipping check for Starkware Sequencer health as it has been less than ${delta} seconds since last check`,
+      )
+      return lastSeenPendingBlock.lastStatus
+    }
+    let pendingBlockParams: GetBlockResponse
+    try {
+      pendingBlockParams = await config.starkwareConfig.provider.getBlock('pending')
+    } catch (e: any) {
+      if (e.providerStatusCode === 504) {
+        Logger.warn(
+          `Request to fetch pending block timed out.  Status Code: ${e.providerStatusCode}.  Sequencer: UNHEALTHY`,
+        )
+        lastSeenPendingBlock.lastStatus = false
+        return false
+      }
+      throw e
+    }
+    const isBatcherHealthy = checkBatcherHealthy(lastSeenPendingBlock, pendingBlockParams)
+    lastSeenPendingBlock = {
+      parentBlockHash: pendingBlockParams.parent_block_hash,
+      txnCount: Object.keys(pendingBlockParams.transactions).length,
+      lastUpdated: currentTime,
+      lastStatus: isBatcherHealthy,
+    }
+    return isBatcherHealthy
+  }
+}
+
+const checkBatcherHealthy = (
+  lastSeenPendingBlock: PendingBlockState,
+  pendingBlockParams: GetBlockResponse,
```

Comment:
Two different types used to compare two of the same concepts (last seen & current pending block) make it hard to read/reason.

```suggestion
  pendingBlockPrevious: GetBlockResponse,
  pendingBlockCurrent: GetBlockResponse,
```

Reply:
These are different.  `PendingBlockState` stores the last status and last updated at timestamp, which used internally by the EA.  I'll rename to make it clearer.

Reply:
I know they are currently different, but you can avoid having that custom type altogether and just use  `PendingBlockState` + in-line calculations.

Slightly more computation for more readability.